contracts/utils/test/lib_address_array.ts

```diff
@@ -0,0 +1,156 @@
+        await blockchainLifecycle.revertAsync();
+    });
+    before(async () => {
+        // Deploy LibAddressArray
+        lib = await TestLibAddressArrayContract.deployFrom0xArtifactAsync(
+            artifacts.TestLibAddressArray,
+            provider,
+            txDefaults,
+        );
+    });
+    beforeEach(async () => {
+        await blockchainLifecycle.startAsync();
+    });
+    afterEach(async () => {
+        await blockchainLifecycle.revertAsync();
+    });
+
+    describe('append', () => {
+        it('should append to empty array', async () => {
+            const addr = addressUtils.generatePseudoRandomAddress();
+            const result = await lib.publicAppend.callAsync([], addr);
+            const expected = [addr];
+            expect(result).to.deep.equal(expected);
+        });
+
+        it('should append to non-empty array', async () => {
+            const arr = _.times(3, () => addressUtils.generatePseudoRandomAddress());
+            const addr = addressUtils.generatePseudoRandomAddress();
+            const expected = [...arr, addr];
+            const result = await lib.publicAppend.callAsync(arr, addr);
+            expect(result).to.deep.equal(expected);
+        });
+
+        it('should revert if the free memory pointer was moved to before the end of the array', async () => {
+            const arr = _.times(3, () => addressUtils.generatePseudoRandomAddress());
+            const addr = addressUtils.generatePseudoRandomAddress();
+            return expectContractCallFailedAsync(
+                lib.testAppendRealloc.callAsync(arr, new BigNumber(-1), addr),
+                RevertReason.InvalidFreeMemoryPtr,
+            );
+        });
+
+        it('should keep the same memory address if free memory pointer does not move', async () => {
+            const arr = _.times(3, () => addressUtils.generatePseudoRandomAddress());
+            const addr = addressUtils.generatePseudoRandomAddress();
+            const expected = [...arr, addr];
+            const [result, oldArrayMemStart, newArrayMemStart] = await lib.testAppendRealloc.callAsync(
+                arr,
+                new BigNumber(0),
+                addr,
+            );
+            expect(result).to.deep.equal(expected);
+            expect(newArrayMemStart).bignumber.to.be.equal(oldArrayMemStart);
+        });
+
+        it('should change memory address if free memory pointer advances', async () => {
+            const arr = _.times(3, () => addressUtils.generatePseudoRandomAddress());
+            const addr = addressUtils.generatePseudoRandomAddress();
+            const expected = [...arr, addr];
+            const [result, oldArrayMemStart, newArrayMemStart] = await lib.testAppendRealloc.callAsync(
+                arr,
+                new BigNumber(1),
+                addr,
+            );
+            expect(result).to.deep.equal(expected);
+            expect(newArrayMemStart).bignumber.to.not.be.equal(oldArrayMemStart);
```

Comment:
To ensure `append` is not overwriting other areas of memory, it could be good to explicitly test for the expected value here. The new array should be placed at the end of memory.